### PR TITLE
fix(seaweedfs): close the 4.31 rename fallout on the 1.5.x→1.6 upgrade path

### PR DESCRIPTION
**Pull request by myasnikovdaniil**
## What this PR does

Cozystack v1.5.0 bumped the vendored SeaweedFS chart 4.0.405 → 4.31.0, which renamed every workload after the Helm release (`<name>-system-*`). StatefulSet names are immutable, so upgrades through 1.5.x stood up a second, empty set beside the running one instead of renaming. #3282 pinned `fullnameOverride: seaweedfs` so 1.6 adopts running workloads in place — this PR closes the remaining holes on that adoption path, found by driving a disposable 3-node cluster through a real v1.4.5 → v1.5.3 → main upgrade with SeaweedFS tenants planted in every reachable state.

- **The naming guard moves into `packages/system/seaweedfs`** — the render a platform upgrade actually re-renders: the `<name>-system` HelmRelease pulls this chart from a platform-managed ExternalArtifact, so the previous guard in `extra/seaweedfs` was never in the path. Upgrading a fresh-1.5.x tenant therefore created an empty chart-named set beside its live data and flipped the guard's own legacy-data signal. `extra/` keeps a sibling copy for operator visibility; a bats parity suite pins the two detection blocks byte-identical.
- **The guard refuses instead of guessing when both naming generations exist.** Nothing durable distinguishes a duplicate that never served from one that served and crashed: claim timestamps invert during the recovery runbook's own re-bind, `readyReplicas: 0` is a snapshot, and Helm birth order answers "which generation is original", not "is the other one empty". Exactly one generation present is decidable, and the render proceeds (or refuses as class S) on its own.
- **Cluster-scoped RBAC is named per namespace again.** 4.31 named four cluster-scoped objects after the release — identical for every tenant — so all tenants collided on one ClusterRole/ClusterRoleBinding and only the last-reconciled tenant's COSI provisioner kept its RBAC. Names return to `global.seaweedfs.serviceAccountName`; three of four are byte-identical to pre-4.31 and adopt in place.
- **The `seaweedfs-db` hand-over runs for every instance name.** Migration 43 compared the owning release against the literal `seaweedfs-system`, so an instance named `foo` was skipped and its CNPG Cluster — the filer metadata for every object in that tenant's S3 — was pruned on the next reconcile, PVC included. The comparison now matches the `-system` suffix (shared `lib/seaweedfs-db-adopt.sh`), and new migration 53 re-runs the hand-over for clusters already past 43, before anything re-renders.
- **`hack/seaweedfs-naming-audit.sh` + `docs/operations/seaweedfs-431-rename-recovery.md`** — what the guard's refusal points operators at: read-only classification (`L` / `S` / `MIXED`, naming the candidate duplicate from relative PV vintage, never a clock window) and the recovery procedures.

**Scope.** The supported SeaweedFS deployment is the tenant module, which hardcodes the instance name (`packages/apps/tenant/templates/seaweedfs.yaml`); a tenant only enables or disables it. The runbook and its selectors are scoped to that name; instances created directly against the API under other names are classified by the audit but routed to escalation. Zone/pool keys of ~40 or more characters fall outside the guard's reconstruction — an accepted limit, recorded in `_naming.tpl` and the runbook.

**Upgrade impact.** 1.4.x → 1.6: no manual steps — one generation, adopted in place. 1.5.x → 1.6 with SeaweedFS: migration 53 protects every `seaweedfs-db` first; then the upgrade **refuses** for any tenant holding both naming generations until the operator resolves the duplicate. This is deliberate — which generation holds the data is not decidable from inside a render, and guessing wrong destroys it. Release notes should present the refusal as expected behavior.

**Testing.** 55 chart unit tests across both packages (including new MultiZone zone-component guard cases — the suite previously had no zone shapes), 11 audit bats, 8 guard-parity bats, migration bats. Validated end-to-end on a disposable 3-node cluster driven v1.4.5 → v1.5.3 → main with five tenants covering: never-saw-4.31 (renders untouched), fresh-1.5.x (refused as class S), wedged duplicate, split duplicate, and a non-default-named instance (its database survives only with this fix). The audit classifies all five correctly, and its two independent signals — revision-1 birth scheme and relative PV vintage — agree on every MIXED tenant.

Related: #3282 (fullnameOverride pin), #3335 (etcd adoption backup gate — separate, also required for the 1.5.x→1.6 path).

### Screenshots

Not a UI change.

### Downstream repositories

- [x] No downstream repository is affected by this change

Walked the trigger map against the diff: no package added/renamed under `packages/{apps,extra}/`, the `packages/core/platform/values.yaml` change is only the migrations `targetVersion` bump (no `spec.components.platform.values.*` key changes), no variant/bundle/component changes, no asset renames, no ApplicationDefinition semantic changes.

### Release note

```release-note
fix(seaweedfs): the 1.6 upgrade no longer renames a SeaweedFS instance away from its data. The naming guard now runs in the chart a platform upgrade actually re-renders and refuses when both pre- and post-4.31 naming generations exist; hack/seaweedfs-naming-audit.sh and docs/operations/seaweedfs-431-rename-recovery.md guide recovery, and the refusal is expected for tenants that passed through 1.5.x. Cluster-scoped COSI RBAC is named per namespace again (the 4.31 release-based names collided across tenants), and the seaweedfs-db hand-over runs for every instance name — previously an instance not named `seaweedfs` had its filer metadata database pruned on upgrade; new migration 53 repairs clusters that already ran the old hand-over.
```

## Summary by CodeRabbit

* **New Features**
  * Added fail-closed upgrade protections for SeaweedFS naming migrations (mixed/damaged classification), including safer behavior when cluster visibility is limited.
  * Added cluster-scoped RBAC uniqueness safeguards to prevent cross-tenant name collisions during rendering.
  * Improved SeaweedFS database adoption/repair migrations and strengthened post-delete cleanup ownership checks.
* **Bug Fixes**
  * Hardened SeaweedFS 4.31 rename recovery and PV rebind flow, including reclaim policy preservation and long/non-default instance-name edge cases.
* **Documentation**
  * Expanded the SeaweedFS 4.31 rename-recovery runbook with clarified auditing, verification, and escalation.
* **Tests**
  * Added/expanded integration and Helm rendering tests for the above scenarios and refusal/fail-closed behavior.

### CODE REVIEW

File: docs/operations/seaweedfs-431-rename-recovery.md [hunk 1]
```diff
@@ -2,56 +2,115 @@
...
+kubectl -n "$ns" apply -f - <<EOF
+apiVersion: v1
+kind: Pod
+metadata:
+  name: swfs-emptiness-check
+spec:
+  restartPolicy: Never
+  containers:
+  - name: inspect
+    image: busybox:1.37
```
[the remark this answers is not in the document]
Comment (myasnikovdaniil) [reply docs/operations/seaweedfs-431-rename-recovery.md:91]: Thanks, but `busybox:1.37` does exist on Docker Hub — it's the `1.37.0` tag (last pushed 2026-05-22), which is newer than `1.36`. `busybox:1.37` pulls fine, so there's no ImagePullBackOff to avoid here and pinning back to `1.36` would move the runbook backwards. Leaving it as is.

File: docs/operations/seaweedfs-431-rename-recovery.md [hunk 2]
```diff
@@ -84,22 +143,35 @@ for pvc in $(kubectl -n "$ns" get pvc -o name | sed 's|persistentvolumeclaim/||'
...
   reclaim=$(kubectl get pv "$pv" -o jsonpath='{.spec.persistentVolumeReclaimPolicy}')
+  if [ -z "$(kubectl get pv "$pv" -o jsonpath='{.metadata.annotations.cozystack\.io/original-reclaim-policy}')" ]; then
+    kubectl annotate pv "$pv" "cozystack.io/original-reclaim-policy=${reclaim}"
+  fi
 
   # Keep the PV (and the data) when the claim goes away.
   kubectl patch pv "$pv" -p '{"spec":{"persistentVolumeReclaimPolicy":"Retain"}}'
   kubectl -n "$ns" delete pvc "$pvc"
   # A Released PV cannot be re-bound until its old claimRef is cleared.
   kubectl patch pv "$pv" --type json -p '[{"op":"remove","path":"/spec/claimRef"}]'
```
[the remark this answers is not in the document]
Comment (myasnikovdaniil) [reply docs/operations/seaweedfs-431-rename-recovery.md:159]: Fixed in 5236af25b. The re-bind loop now gates the `delete pvc` on both recording the original reclaim policy and the `patch … Retain` succeeding (`|| { echo …; break; }`), so with no `set -e` a failed patch aborts the loop instead of deleting the PVC while its PV is still Delete. The Step 5 restore loop chains the annotation removal after the restore patch with `&&` (see the thread below).

File: docs/operations/seaweedfs-431-rename-recovery.md [hunk 3]
```diff
@@ -195,3 +369,16 @@ kubectl -n "$ns" get endpoints seaweedfs-s3
...
+Only once that is true, restore the reclaim policy Step 2 stashed on each PV. Until this runs the volumes are `Retain`, which is deliberate: it is what makes an accidental claim deletion during recovery survivable.
+
+```sh
+ns=<tenant>
+for pvc in $(kubectl -n "$ns" get pvc -o name | sed 's|persistentvolumeclaim/||' | grep -E '^data1-seaweedfs-volume'); do
+  pv=$(kubectl -n "$ns" get pvc "$pvc" -o jsonpath='{.spec.volumeName}')
+  orig=$(kubectl get pv "$pv" -o jsonpath='{.metadata.annotations.cozystack\.io/original-reclaim-policy}')
+  [ -n "$orig" ] || continue
+  kubectl patch pv "$pv" -p "{\"spec\":{\"persistentVolumeReclaimPolicy\":\"${orig}\"}}"
+  kubectl annotate pv "$pv" cozystack.io/original-reclaim-policy-
```
[the remark this answers is not in the document]
Comment (myasnikovdaniil) [reply docs/operations/seaweedfs-431-rename-recovery.md]: Fixed in 5236af25b: the restore now runs `kubectl patch … && kubectl annotate …-`, so the saved `cozystack.io/original-reclaim-policy` annotation is dropped only after the reclaim-policy patch succeeds and is preserved for a retry if the patch fails.

File: packages/extra/seaweedfs/templates/hooks/cleanup.yaml [hunk 4]
```diff
@@ -141,6 +164,38 @@ spec:
...
+              # so confirm THIS release owns it before deleting. A Client instance
+              # renders no <name>-db and shares another app's filer; a second
+              # Simple instance in the same namespace loses Helm's ownership race
+              # but still installs. Deleting on name alone would take a different
+              # app's database.
+              echo "Deleting SeaweedFS filer-metadata database for {{ .Release.Name }}..."
+              owner=$(kubectl get cluster.postgresql.cnpg.io -n {{ .Release.Namespace }} seaweedfs-db \
+                        -o jsonpath='{.metadata.annotations.meta\.helm\.sh/release-name}' 2>/dev/null || true)
+              if [ -z "$owner" ]; then
+                echo "No Cluster/seaweedfs-db owned by a Helm release here; nothing to reclaim."
```
[the remark this answers is not in the document]
Comment (myasnikovdaniil) [reply packages/extra/seaweedfs/templates/hooks/cleanup.yaml]: Fixed in 2f7ec784c. The ownership read no longer uses `|| true`; it captures stderr and treats only a genuine NotFound as absence — any other error (Forbidden, timeout, API failure) prints the message and `exit 1`s so the Job retries, matching the fail-closed handling the delete path already had. Added a regression assertion in `cleanup_test.yaml` that the lookup no longer swallows failures.

File: packages/extra/seaweedfs/tests/fullname_override_test.yaml [hunk 5]
```diff
@@ -4,16 +4,24 @@ suite: seaweedfs legacy-naming guard
...
-# 1.5.x (only the renamed volumes exist → PV re-bind, Step 2), and a D-split
-# tenant (both sets present AND the renamed volume servers are live → reconcile
+# Three states stop the render instead of adopting: a tenant installed FRESH on
+# 1.5.x (only the renamed volumes exist → PV re-bind, Step 2), a tenant an
+# unguarded upgrade already damaged (BOTH generations exist but the renamed
+# volumes are OLDER — the data was born there and the chart-named set is the
+# empty one → Step 2a + Step 2), and a D-split tenant (both sets present, the
+# legacy volumes are older, AND the renamed volume servers are live → reconcile
 # the split, Step 3). A duplicate that never served (D-wedged, zero ready
 # replicas) renders through and is adopted.
```
[the remark this answers is not in the document]
Comment (myasnikovdaniil) [reply packages/extra/seaweedfs/tests/fullname_override_test.yaml]: Fixed in 10af926db. The stale claim lived at `templates/seaweedfs.yaml` L110/L119-120 and the `fullname_override_test.yaml` header — both now say a D-wedged duplicate (zero ready replicas) is refused too, since readyReplicas is a snapshot and can't be told apart from D-split. Comments only, no classifier change. The seaweedfs.yaml note at L137-142 and the `system/seaweedfs` naming-guard already described the presence-only refusal, so those were left as is.

**Comment by IvanHunters** [comment]
> **Updated** — this replaces my earlier verdict on this comment. The first pass reasoned about the migration shell portability but did not run the shipped test under the CI shell; re-running with that check flips the verdict.

## Verdict

NOT LGTM

One MAJOR regression, confirmed by this PR's own red `Unit & controller tests` job. Everything else verified clean (upgrade path, config-toggle corners, per-namespace RBAC, cleanup hook, guard parity, vendored-patch wiring).

## Findings

**[MAJOR]** `packages/core/platform/images/migrations/migrations/43:23` and `…/53:41` — migration bashism under `#!/bin/sh` breaks its own test under dash

Both migrations start with `#!/bin/sh` + `set -euo pipefail`, and the new test `hack/migration-seaweedfs-db-adopt.bats` runs them as `sh "$MIG_DIR/43"` / `sh "$MIG_DIR/53"`. On the GitHub `ubuntu` runner `/bin/sh` is **dash**, which does not implement `pipefail`: the script aborts on line 1 with `set: Illegal option -o pipefail` (exit 2) before doing any work. The test expects `rc=0` and gets `rc=2`, so the `Unit & controller tests` job is RED (`migrations/43: 23: set: Illegal option -o pipefail`).

Production is unaffected — the migrations image is `FROM alpine`, so `/bin/sh` is busybox ash, which supports `pipefail`. That is exactly why checking only the runtime image misses this; the failing environment is the test's dash, not production's ash.

Fix (any one):
- give both migrations `#!/bin/bash` **and** exec them via their shebang in the test (`"$MIG_DIR/43"`, not `sh "$MIG_DIR/43"`) — this is what migration `50` already does; or
- make the scripts dash-clean (drop `pipefail`); or
- run them in the test under the production interpreter (`bash` / busybox `ash`), never the runner's `sh`.

Repro without CI: `dash packages/core/platform/images/migrations/migrations/43`.

## Caveats (verified, non-blocking)

- **Upgrade path**: migration `53` is strictly above the current highest (`52`), `targetVersion` bumped `53`→`54` (no silent skip, not rebase-stale); `lib/seaweedfs-db-adopt.sh` is idempotent, fails closed on any non-absent kubectl error, and matches the `-system` suffix for every instance name. Correct — modulo the dash bug above, which blocks the tests that would prove it.
- **Config-toggle matrix (rendered)**: `Simple` / `MultiZone` emit `<name>-db` + `<name>-system` with `dependsOn: seaweedfs-db` satisfied; `Client` emits zero HelmReleases (both are inside the same `{{- if not (eq .Values.topology "Client") }}` gate at `seaweedfs.yaml:95` and `seaweedfs-db.yaml:1`), so there is no dangling `dependsOn`.
- **Cluster-scoped RBAC** renders per-namespace-unique (`<ns>-seaweedfs-*`) via the `seaweedfs.compat` shim; regression test + runtime guard present. The `charts/` edits are the correct shape: `patches/cluster-scoped-names-per-namespace.patch` exists and the Makefile `update` target applies it (no silent revert on `make update`).
- **Cleanup Job**: `activeDeadlineSeconds` + `--wait=false` + ownership check + release-scoped selector, PSS-restricted. `guard-parity.bats` pins the two detection blocks byte-identical.
- NIT: `packages/system/seaweedfs/values.yaml:7` default `serviceAccountName: "tenant-foo-seaweedfs"` reads like a leaked concrete tenant name (harmless — overridden per namespace, and the guard fails closed if it were ever effective).

**Comment by myasnikovdaniil** [comment]
@IvanHunters Fixed in a38843c5 — thanks, the repro was exact.

Rather than move the shebang or drop pipefail, I took your third option and made the test run the migrations the way `run-migrations.sh` does: by path, inside the image's own pinned alpine base, read out of the migrations Dockerfile so the interpreter under test can't drift from the one they ship on. `pipefail` turned out to be load-bearing rather than decorative — `lib/cozystack-version.sh` pipes the rendered manifest into `kubectl apply`, so without it a failed render stamps the version from empty input — which is what ruled out the dash-clean variant. Note this makes the unit-test job depend on docker; there's deliberately no host-shell fallback, since every fallback available is a shell production never uses.

One thing the switch surfaced: `cozytest.sh`'s awk generator rewrites every bare `}` in column 0 into `return 0` + `}`, so a helper that falls off its own end returns 0 regardless of what it ran. The first version of `run_migration` did exactly that — the three fail-closed tests went green while the container was exiting 1. It now captures and returns the status by hand.

15/15 pass. Verified the interpreter under test is BusyBox v1.37.0 ash with pipefail supported, and that a fail-open mutation in `lib/seaweedfs-db-adopt.sh` turns the suite red.

On the NIT — `packages/system/seaweedfs/values.yaml:7` predates this PR and isn't in its diff, so I've left it out rather than widen the change. Happy to take it separately if you'd like.

**Review by IvanHunters — APPROVED** [review]
## Verdict

LGTM with non-blocking notes

`fix(seaweedfs)` closes the 4.31 rename fallout on the 1.5.x→1.6 upgrade path. Zero CRITICAL / MAJOR / MINOR blocking findings. The one cozy-invariant that fired (`charts-direct-edit`, MINOR) is fully neutralised by the correct form. Each critical node was verified by execution (helm unittest, helm template renders, migration bats in busybox ash via docker, guard-parity/audit bats, canary), not just by reading.

## Verified by execution

- **Migrations 43 (edit) / 53 (new) + targetVersion.** main is at 52 (target 53); this PR adds 53 and bumps target to 54 (strictly greater), so `run-migrations.sh seq CURRENT 53` genuinely re-runs the hand-over on already-upgraded clients. `hack/check-migrations-target.sh` confirms. Migration bats (`hack/migration-seaweedfs-db-adopt.bats`) 5/5 green on the key paths (default / non-default handover, mixed fleet, protect-missing-keep, repair-53); full 15-test set passes in the "Unit & controller tests" CI job.
- **Vendored-chart invariant handled correctly.** Edits under `packages/system/seaweedfs/charts/seaweedfs/templates/{cosi/cosi-cluster-role.yaml,shared/cluster-role.yaml}` are captured in `packages/system/seaweedfs/patches/cluster-scoped-names-per-namespace.patch`, and `packages/system/seaweedfs/Makefile:update` re-applies it (`patch -p4 < patches/…`) — no silent revert on `make update`.
- **Cluster-scoped RBAC de-collision.** `extra/seaweedfs/templates/seaweedfs.yaml:226-228` sets `global.serviceAccountName: "{{ .Release.Namespace }}-seaweedfs"` on the emitted `<name>-system` HR; `seaweedfs.compat` folds it to the canonical key; `cluster-scoped-rbac-guard.yaml` asserts `$sa == <namespace>-seaweedfs`; `tests/cluster_scoped_names_test.yaml` confirms name uniqueness across two namespaces. The guard render does not fail.
- **Shell-portability trap did not fire.** Migrations 43/53 are `#!/bin/sh` + `set -euo pipefail`; both production and the bats test run them through busybox ash (`docker run alpine … /migrations/<n>` by path, not `sh <file>` on a dash runner). CI green confirms.
- **Naming guard is fail-closed.** Local render: fresh install renders normally, client-side/unittest is skipped (empty canary), the blind-upgrade canary (`make test-guard-fail-closed`) refuses as intended; class-S/MIXED is a deliberate refuse. Guard-parity bats confirms byte-identical detection blocks between `system` and `extra`.
- **Cleanup hook is safe** (`extra/seaweedfs/templates/hooks/cleanup.yaml`): deletion of `Cluster/seaweedfs-db` is gated on `topology != Client` + RBAC `resourceNames:[seaweedfs-db]` + a runtime ownership recheck (fail-closed on foreign owner); PVCs by release-scoped label; `--wait=false` throughout; `activeDeadlineSeconds: 120`, `backoffLimit: 3`, and hook-delete-policy without `hook-failed` (a failed Job is kept for debugging).

## Caveats (non-blocking)

1. **E2E upgrade not independently reproduced** (hermetic review, no live cluster; the PR's "E2E Tests" job was still pending). Logic verified via renders and migration bats, but the real N-1 → N run rests on the author's report plus the pending CI job.
2. **Deliberate refuse for class-S/MIXED tenants:** their `<name>-system` HR (`remediation.retries: -1`) will loop-fail until a manual duplicate resolution per the runbook (`docs/operations/seaweedfs-431-rename-recovery.md`). This is a documented safe trade-off (the alternative is data loss); the release note warns about it.
3. This is the first PR to introduce a docker-daemon dependency in the "Unit & controller tests" job (migration bats). Confirmed green on the current runner, but it depends on docker being present on future runners.
